### PR TITLE
build: Enhance Codespaces

### PR DESCRIPTION
**Pull request by alexec**
This PR:

- Makes `make argoexec-image` run in just a few seconds when rebuilding from cache. The `.git` folder gets modified every time you run git. This mean that the Docker caches being invalidated. `GIT_COMMIT` etc must be passed as Docker build .args
- Codespaces is not very polished. E.g. the "universal" image is not "universal" because it does not work on Apple Silicon.
- I've pinned Node to 16. While both 16 and 18 are LTS, 18 has the legacy SSL issue, to which there does not appear to be any simple fix.

If pre-built, you can start a Codespace instantly. When not pre-built, it takes several minutes.

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
+      "version": "16"
+    },
+    "ghcr.io/devcontainers/features/docker-in-docker:2": {}
+  },
+  "hostRequirements": {
+    "cpus": 4
+  },
+  "onCreateCommand": ".devcontainer/pre-build.sh",
+  "workspaceMount": "source=${localWorkspaceFolder},target=/home/vscode/go/src/github.com/argoproj/argo-workflows,type=bind",
+  "workspaceFolder": "/home/vscode/go/src/github.com/argoproj/argo-workflows"
```
Comment (tczhao) [inline .devcontainer/devcontainer.json:17]: Could we keep using the `${localEnv:HOME}` instead of `/home/vscode` so it still works on local devcontainer

**Review by tczhao — CHANGES_REQUESTED** [review]
Perhaps we need to add some info about the `kit up` in 
https://github.com/tczhao/argo/blob/master/docs/running-locally.md

**Review by terrytangyuan — APPROVED** [review]
LGTM. Please take a look at the comment above.

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (alexec) [reply .devcontainer/devcontainer.json:17]: Let me test. I was fixing something by doing this.

**Review by sarabala1979 — APPROVED** [review]

**Comment by alexec** [comment]
I'll fix vscode problem if needed later.

**Pull request merged**